render/vulkan/vulkan.c

```diff
@@ -0,0 +1,596 @@
+			&avail_extc, avail_ext_props);
+		if (res != VK_SUCCESS) {
+			wlr_vk_error("  Could not enumerate device extensions", res);
+			continue;
+		}
+
+		const char *name = VK_EXT_PCI_BUS_INFO_EXTENSION_NAME;
+
+		if (find_extensions(avail_ext_props, avail_extc, &name, 1) != NULL) {
+			wlr_log(WLR_INFO, "  Can't check pci info of device, "
+				"as it does not support the PCI_BUS_INFO extensions");
+			continue;
+		}
+
+		VkPhysicalDevicePCIBusInfoPropertiesEXT bus_info = {0};
+		bus_info.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PCI_BUS_INFO_PROPERTIES_EXT;
+
+		VkPhysicalDeviceProperties2 props = {0};
+		props.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROPERTIES_2;
+
+		props.pNext = &bus_info;
+
+		ini->api.getPhysicalDeviceProperties2(phdev, &props);
+
+		wlr_log(WLR_INFO, "  PCI bus: %04x:%02x:%02x.%x", bus_info.pciDomain,
+			bus_info.pciBus, bus_info.pciDevice, bus_info.pciFunction);
+
+		if (bus_info.pciDevice == drm_pci_info.dev &&
+				bus_info.pciBus == drm_pci_info.bus &&
+				bus_info.pciDomain == drm_pci_info.domain &&
+				bus_info.pciFunction == drm_pci_info.func) {
+			wlr_log(WLR_INFO, "Found matching vulkan device from drm pci info: %s",
+				phdev_props.deviceName);
+			return phdev;
+		}
+	}
+
+	return VK_NULL_HANDLE;
+}
+
+struct wlr_vk_device *vulkan_device_create(struct wlr_vk_instance *ini,
+		VkPhysicalDevice phdev, size_t ext_count, const char **exts) {
+	VkResult res;
+
+	// check for extensions
+	uint32_t avail_extc = 0;
+	res = vkEnumerateDeviceExtensionProperties(phdev, NULL,
+		&avail_extc, NULL);
+	if (res != VK_SUCCESS || avail_extc == 0) {
+		wlr_vk_error("Could not enumerate device extensions (1)", res);
+		return NULL;
+	}
+
+	VkExtensionProperties avail_ext_props[avail_extc + 1];
+	res = vkEnumerateDeviceExtensionProperties(phdev, NULL,
+		&avail_extc, avail_ext_props);
+	if (res != VK_SUCCESS) {
+		wlr_vk_error("Could not enumerate device extensions (2)", res);
+		return NULL;
+	}
+
+	for (size_t j = 0; j < avail_extc; ++j) {
+		wlr_log(WLR_DEBUG, "Vulkan device extension %s v%"PRIu32,
+			avail_ext_props[j].extensionName, avail_ext_props[j].specVersion);
+	}
+
+	// create device
+	struct wlr_vk_device *dev = calloc(1, sizeof(*dev));
+	if (!dev) {
+		wlr_log_errno(WLR_ERROR, "allocation failed");
+		return NULL;
+	}
+
+	dev->phdev = phdev;
+	dev->instance = ini;
+	dev->drm_fd = -1;
+	dev->extensions = calloc(16 + ext_count, sizeof(*ini->extensions));
+	if (!dev->extensions) {
+		wlr_log_errno(WLR_ERROR, "allocation failed");
+		goto error;
+	}
+
+	// find extensions
+	for (unsigned i = 0; i < ext_count; ++i) {
+		if (find_extensions(avail_ext_props, avail_extc, &exts[i], 1)) {
+			wlr_log(WLR_DEBUG, "vulkan device extension %s not found",
+				exts[i]);
+			continue;
+		}
+
+		dev->extensions[dev->extension_count++] = exts[i];
+	}
+
+	// for dmabuf importing we require at least the
+	// 'external_memory_fd, external_memory_dma_buf, queue_family_foreign'
```

Comment:
Please consider requiring the modifier extension as well. Everyone has started to implement it already, and those that haven't are guaranteed to have implementations buggy enough to fall back to OpenGL.

Reply:
It's not just nice to have, it's *required* to import dmabufs. Doing so without the modifiers extension is undefined behavior. (You also use modifier structs in your code)

Reply:
It's set to linear in another section IIRC. Which only works (worked-ish?) on Intel GPUs under specific conditions. Like mentioned, you really do want to require the modifier extension, otherwise it's undefined behaviour.

Reply:
We actually require them below. This comment is an old leftover.